File: password_forge.js
// password_forge.js
#!/usr/bin/env node
'use strict';

const crypto = require('crypto');
const fs = require('fs');

// ANSI colors
const COLORS = {
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    red: '\x1b[91m',
    blue: '\x1b[94m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

// Встроенный словарь (топ-1000 урезанный)
const commonPasswords = new Set([
    'password', '123456', '12345678', '1234', 'qwerty', '12345', 'dragon',
    'baseball', 'football', 'letmein', 'monkey', 'abc123', 'mustang', 'michael',
    'shadow', 'master', 'jennifer', '111111', '2000', 'jordan', 'superman',
    'harley', '1234567', 'fuckyou', 'hunter', 'trustno1', 'ranger', 'buster',
    'thomas', 'tigger', 'robert', 'soccer', 'battle', 'killer', 'hockey', 'george',
    'andrew', 'charlie', 'donald', 'asshole', 'qwertyuiop', 'liverpool', 'nicole',
    'daniel', 'ginger', 'whatever', 'samantha', 'jasmine', 'passw0rd', 'qazwsx'
]);

function isCommon(pwd) {
    return commonPasswords.has(pwd.toLowerCase());
}

function estimateEntropy(pwd, charsetSize) {
    if (!charsetSize) {
        const unique = new Set(pwd);
        charsetSize = unique.size;
    }
    if (charsetSize <= 1) return 0;
    return pwd.length * Math.log2(charsetSize);
}

function formatTime(seconds) {
    if (!isFinite(seconds)) return '∞ (слишком долго)';
    if (seconds < 1) return seconds.toFixed(3) + ' секунд';
    if (seconds < 60) return Math.round(seconds) + ' секунд';
    if (seconds < 3600) return Math.round(seconds/60) + ' минут';
    if (seconds < 86400) return Math.round(seconds/3600) + ' часов';
    if (seconds < 31536000) return Math.round(seconds/86400) + ' дней';
    return Math.round(seconds/31536000) + ' лет';
}

function generatePronounceable(length) {
    const vowels = 'aeiouy';
    const consonants = 'bcdfghjklmnpqrstvwxz';
    let pwd = '';
    for (let i = 0; i < length; i++) {
        if (i % 2 === 0) pwd += consonants[Math.floor(Math.random() * consonants.length)];
        else pwd += vowels[Math.floor(Math.random() * vowels.length)];
    }
    if (Math.random() < 0.5) pwd = pwd[0].toUpperCase() + pwd.slice(1);
    return pwd;
}

function generatePassword(length, useUpper, useLower, useDigits, useSpecial,
                          excludeAmbiguous, pronounceable) {
    if (pronounceable) return generatePronounceable(length);

    let chars = '';
    if (useLower) chars += 'abcdefghijklmnopqrstuvwxyz';
    if (useUpper) chars += 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    if (useDigits) chars += '0123456789';
    if (useSpecial) chars += '!\"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~';
    if (!chars) throw new Error('Ни один набор символов не выбран');

    if (excludeAmbiguous) {
        const ambiguous = 'O0o1lI';
        chars = chars.split('').filter(c => !ambiguous.includes(c)).join('');
    }

    let pwd = '';
    for (let i = 0; i < length; i++) {
        pwd += chars[Math.floor(Math.random() * chars.length)];
    }
    return pwd;
}

function parseArgs() {
    const args = process.argv.slice(2);
    const opts = {
        length: 16,
        count: 1,
        uppercase: true,
        lowercase: true,
        digits: true,
        special: true,
        exclude: false,
        pronounce: false,
        minEntropy: 0,
        quiet: false,
        output: '',
        help: false
    };
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        switch (arg) {
            case '-l':
            case '--length': opts.length = parseInt(args[++i]); break;
            case '-c':
            case '--count': opts.count = parseInt(args[++i]); break;
            case '-u':
            case '--uppercase': opts.uppercase = true; break;
            case '-L':
            case '--lowercase': opts.lowercase = true; break;
            case '-d':
            case '--digits': opts.digits = true; break;
            case '-s':
            case '--special': opts.special = true; break;
            case '-e':
            case '--exclude': opts.exclude = true; break;
            case '-p':
            case '--pronounce': opts.pronounce = true; break;
            case '-m':
            case '--min-entropy': opts.minEntropy = parseFloat(args[++i]); break;
            case '-q':
            case '--quiet': opts.quiet = true; break;
            case '-o':
            case '--output': opts.output = args[++i]; break;
            case '-h':
            case '--help': opts.help = true; break;
            default: console.error(colorize(`Неизвестная опция: ${arg}`, 'red')); process.exit(1);
        }
    }
    return opts;
}

function main() {
    const opts = parseArgs();
    if (opts.help) {
        console.log(`Usage: node password_forge.js [options]
Options:
  -l, --length <n>       Длина пароля (default 16)
  -c, --count <n>        Количество паролей (default 1)
  -u, --uppercase        Включать заглавные (default on)
  -L, --lowercase        Включать строчные (default on)
  -d, --digits           Включать цифры (default on)
  -s, --special          Включать спецсимволы (default on)
  -e, --exclude          Исключать неоднозначные символы
  -p, --pronounce        Генерировать произносимые пароли
  -m, --min-entropy <n>  Минимальная энтропия (бит)
  -q, --quiet            Не выводить оценку сложности
  -o, --output <file>    Сохранить в файл
  -h, --help             Справка`);
        return;
    }

    const passwords = [];
    for (let i = 0; i < opts.count; i++) {
        let pwd, attempts = 0;
        do {
            pwd = generatePassword(
                opts.length, opts.uppercase, opts.lowercase, opts.digits, opts.special,
                opts.exclude, opts.pronounce
            );
            attempts++;
            if (attempts > 1000) throw new Error('Не удалось сгенерировать пароль с нужной энтропией');
        } while (estimateEntropy(pwd) < opts.minEntropy);
        passwords.push(pwd);
    }

    let output = '';
    for (const pwd of passwords) {
        const entropy = estimateEntropy(pwd);
        const common = isCommon(pwd);
        const crackSec = entropy === 0 ? Infinity : Math.pow(2, entropy) / 1e9;
        const crackStr = formatTime(crackSec);
        let strength;
        if (entropy < 30) strength = colorize('Слабый', 'red');
        else if (entropy < 50) strength = colorize('Средний', 'yellow');
        else strength = colorize('Сильный', 'green');

        output += `Пароль: ${pwd}\n`;
        if (!opts.quiet) {
            output += `  Энтропия: ${Math.round(entropy)} бит, Стойкость: ${strength}`;
            output += `, Время взлома: ${crackStr}`;
            if (common) output += colorize(' (Словарный!)', 'red');
            output += '\n\n';
        } else {
            output += '\n';
        }
    }

    if (opts.output) {
        try {
            fs.writeFileSync(opts.output, output, 'utf8');
            console.log(colorize(`Результат сохранён в ${opts.output}`, 'green'));
        } catch (err) {
            console.error(colorize(`Ошибка записи файла: ${err.message}`, 'red'));
            process.exit(1);
        }
    } else {
        console.log(output);
    }
}

main().catch(err => {
    console.error(colorize(`Ошибка: ${err.message}`, 'red'));
    process.exit(1);
});

File: password_forge.rb
#!/usr/bin/env ruby
# password_forge.rb
# encoding: UTF-8

require 'optparse'
require 'securerandom'
require 'set'

# ANSI colors
COLORS = {
  green: "\e[92m",
  yellow: "\e[93m",
  red: "\e[91m",
  blue: "\e[94m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

# Встроенный словарь
COMMON_PASSWORDS = %w[
  password 123456 12345678 1234 qwerty 12345 dragon baseball football letmein
  monkey abc123 mustang michael shadow master jennifer 111111 2000 jordan
  superman harley 1234567 fuckyou hunter trustno1 ranger buster thomas tigger
  robert soccer battle killer hockey george andrew charlie donald asshole
  qwertyuiop liverpool nicole daniel ginger whatever samantha jasmine passw0rd qazwsx
].to_set

def is_common?(pwd)
  COMMON_PASSWORDS.include?(pwd.downcase)
end

def estimate_entropy(pwd, charset_size = nil)
  if charset_size.nil?
    charset_size = pwd.chars.to_set.size
  end
  return 0 if charset_size <= 1
  pwd.length * Math.log2(charset_size)
end

def format_time(seconds)
  return "∞ (слишком долго)" if seconds.infinite?
  if seconds < 1
    "#{seconds.round(3)} секунд"
  elsif seconds < 60
    "#{seconds.round} секунд"
  elsif seconds < 3600
    "#{(seconds/60).round} минут"
  elsif seconds < 86400
    "#{(seconds/3600).round} часов"
  elsif seconds < 31536000
    "#{(seconds/86400).round} дней"
  else
    "#{(seconds/31536000).round} лет"
  end
end

def generate_pronounceable(length)
  vowels = 'aeiouy'
  consonants = 'bcdfghjklmnpqrstvwxz'
  pwd = (0...length).map do |i|
    if i.even?
      consonants[rand(consonants.length)]
    else
      vowels[rand(vowels.length)]
    end
  end.join
  if rand < 0.5 && length > 0
    pwd[0] = pwd[0].upcase
  end
  pwd
end

def generate_password(length, use_upper, use_lower, use_digits, use_special,
                      exclude_ambiguous, pronounceable)
  if pronounceable
    return generate_pronounceable(length)
  end

  chars = ''
  chars << 'abcdefghijklmnopqrstuvwxyz' if use_lower
  chars << 'ABCDEFGHIJKLMNOPQRSTUVWXYZ' if use_upper
  chars << '0123456789' if use_digits
  chars << '!\"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~' if use_special
  raise "Ни один набор символов не выбран" if chars.empty?

  if exclude_ambiguous
    ambiguous = 'O0o1lI'
    chars = chars.chars.reject { |c| ambiguous.include?(c) }.join
  end

  (0...length).map { chars[rand(chars.length)] }.join
end

options = {
  length: 16,
  count: 1,
  uppercase: true,
  lowercase: true,
  digits: true,
  special: true,
  exclude: false,
  pronounce: false,
  min_entropy: 0,
  quiet: false,
  output: nil,
  help: false
}

OptionParser.new do |opts|
  opts.banner = "Usage: password_forge.rb [options]"
  opts.on('-l', '--length N', Integer, "Длина пароля (default 16)") { |v| options[:length] = v }
  opts.on('-c', '--count N', Integer, "Количество паролей (default 1)") { |v| options[:count] = v }
  opts.on('-u', '--uppercase', "Включать заглавные (default on)") { |v| options[:uppercase] = true }
  opts.on('-L', '--lowercase', "Включать строчные (default on)") { |v| options[:lowercase] = true }
  opts.on('-d', '--digits', "Включать цифры (default on)") { |v| options[:digits] = true }
  opts.on('-s', '--special', "Включать спецсимволы (default on)") { |v| options[:special] = true }
  opts.on('-e', '--exclude', "Исключать неоднозначные символы") { options[:exclude] = true }
  opts.on('-p', '--pronounce', "Генерировать произносимые пароли") { options[:pronounce] = true }
  opts.on('-m', '--min-entropy N', Float, "Минимальная энтропия (бит)") { |v| options[:min_entropy] = v }
  opts.on('-q', '--quiet', "Не выводить оценку сложности") { options[:quiet] = true }
  opts.on('-o', '--output FILE', "Сохранить в файл") { |v| options[:output] = v }
  opts.on('-h', '--help', "Справка") { options[:help] = true }
end.parse!

if options[:help]
  puts <<~HELP
    Usage: password_forge.rb [options]
    Options:
      -l, --length N       Длина пароля (default 16)
      -c, --count N        Количество паролей (default 1)
      -u, --uppercase      Включать заглавные (default on)
      -L, --lowercase      Включать строчные (default on)
      -d, --digits         Включать цифры (default on)
      -s, --special        Включать спецсимволы (default on)
      -e, --exclude        Исключать неоднозначные символы
      -p, --pronounce      Генерировать произносимые пароли
      -m, --min-entropy N  Минимальная энтропия (бит)
      -q, --quiet          Не выводить оценку сложности
      -o, --output FILE    Сохранить в файл
      -h, --help           Справка
  HELP
  exit
end

passwords = []
options[:count].times do
  attempts = 0
  loop do
    pwd = generate_password(
      options[:length],
      options[:uppercase],
      options[:lowercase],
      options[:digits],
      options[:special],
      options[:exclude],
      options[:pronounce]
    )
    attempts += 1
    raise "Не удалось сгенерировать пароль с нужной энтропией" if attempts > 1000
    if estimate_entropy(pwd) >= options[:min_entropy]
      passwords << pwd
      break
    end
  end
end

output = ''
passwords.each do |pwd|
  entropy = estimate_entropy(pwd)
  common = is_common?(pwd)
  crack_sec = entropy.zero? ? Float::INFINITY : (2 ** entropy) / 1e9
  crack_str = format_time(crack_sec)
  strength = if entropy < 30
               colorize("Слабый", :red)
             elsif entropy < 50
               colorize("Средний", :yellow)
             else
               colorize("Сильный", :green)
             end

  output << "Пароль: #{pwd}\n"
  unless options[:quiet]
    output << "  Энтропия: #{entropy.round} бит, Стойкость: #{strength}"
    output << ", Время взлома: #{crack_str}"
    output << colorize(" (Словарный!)", :red) if common
    output << "\n\n"
  else
    output << "\n"
  end
end

if options[:output]
  begin
    File.write(options[:output], output, encoding: 'UTF-8')
    puts colorize("Результат сохранён в #{options[:output]}", :green)
  rescue => e
    puts colorize("Ошибка записи файла: #{e.message}", :red)
    exit 1
  end
else
  print output
end
